src/client/datascience/delayedFutureExecute.ts

```diff
@@ -0,0 +1,177 @@
+import type { Kernel, KernelMessage } from '@jupyterlab/services';
+import { createDeferred } from '../common/utils/async';
+import { CancellationError } from '../common/cancellation';
+import { traceInfoIfCI } from '../common/logger';
+
+// Wraps a future so that a requestExecute on a session will wait for the previous future to finish before actually executing
+export class DelayedFutureExecute
+    implements Kernel.IShellFuture<KernelMessage.IExecuteRequestMsg, KernelMessage.IExecuteReplyMsg> {
+    private doneDeferred = createDeferred<KernelMessage.IExecuteReplyMsg>();
+    private requestFuture:
+        | Kernel.IShellFuture<KernelMessage.IExecuteRequestMsg, KernelMessage.IExecuteReplyMsg>
+        | undefined;
+    private pendingOnReply: ((msg: KernelMessage.IExecuteReplyMsg) => void | PromiseLike<void>) | undefined;
+    private pendingOnIOPub:
+        | ((msg: KernelMessage.IIOPubMessage<KernelMessage.IOPubMessageType>) => void | PromiseLike<void>)
+        | undefined;
+    private pendingOnStdin:
+        | ((msg: KernelMessage.IStdinMessage<KernelMessage.StdinMessageType>) => void | PromiseLike<void>)
+        | undefined;
+    private pendingHooks: ((
+        msg: KernelMessage.IIOPubMessage<KernelMessage.IOPubMessageType>
+    ) => boolean | PromiseLike<boolean>)[] = [];
+    private pendingInputReplies: (
+        | KernelMessage.IReplyErrorContent
+        | KernelMessage.IReplyAbortContent
+        | KernelMessage.IInputReply
+    )[] = [];
+    private disposed = false;
+    private statusChangedHandler: (_session: Kernel.IKernelConnection, status: KernelMessage.Status) => void;
+    constructor(
+        private kernelConnection: Kernel.IKernelConnection,
+        previousLink: Kernel.IShellFuture<KernelMessage.IExecuteRequestMsg, KernelMessage.IExecuteReplyMsg>,
+        private content: KernelMessage.IExecuteRequestMsg['content'],
+        private disposeOnDone?: boolean,
+        private metadata?: JSONObject
+    ) {
+        // Setup our request based on the previous link finishing
+        if (previousLink) {
+            previousLink.done.then(() => this.requestExecute()).catch((e) => this.doneDeferred.reject(e));
+        }
+
+        // If the kernel dies, finish our future
+        this.statusChangedHandler = (_session: Kernel.IKernelConnection, status: KernelMessage.Status) => {
+            if (status === 'unknown' || status === 'restarting' || status === 'dead') {
+                this.doneDeferred.reject(new CancellationError());
+            }
+        };
+        kernelConnection.statusChanged?.connect(this.statusChangedHandler);
+
+        // Run the handler now to check
+        this.statusChangedHandler(kernelConnection, kernelConnection.status);
+    }
+    public get msg(): KernelMessage.IExecuteRequestMsg {
+        if (this.requestFuture) {
+            return this.requestFuture.msg;
+        }
+        return {
+            content: this.content,
```

Comment:
Just to check my understanding, this path (where the requestFuture is not define) is basically just a dummy placeholder until the previous execute finished right?

Reply:
Yes that's correct. I don't think any of our code ever reads 'msg' anyway.

Reply:
Sounds good, the values for it just didn't quite make sense, so just wanted to confirm that it was a dummy value (essentially).